src/BucketTesting/Validation/CampaignUtilizationValidator.php

```diff
@@ -0,0 +1,137 @@
+
+namespace WMDE\Fundraising\Frontend\BucketTesting\Validation;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\NodeFinder;
+use PhpParser\ParserFactory;
+use Symfony\Component\Filesystem\Exception\FileNotFoundException;
+use WMDE\Fundraising\Frontend\BucketTesting\Campaign;
+use WMDE\Fundraising\Frontend\BucketTesting\CampaignCollection;
+
+/**
+ * @license GNU GPL v2+
+ */
+class CampaignUtilizationValidator {
+
+	/** @see \WMDE\Fundraising\Frontend\Factories\ChoiceFactory */
+	const CHOICE_FACTORY_LOCATION = 'src/Factories/ChoiceFactory.php';
+	/** @see \WMDE\Fundraising\Frontend\BucketTesting\FeatureToggle::featureIsActive */
+	const FEATURE_TOGGLE_METHOD_NAME = 'featureIsActive';
+
+	private $hasValidated = false;
+	private $campaignCollection;
+	private $errorLogger;
+	private $ignoredBuckets = [];
+
+	public function __construct( CampaignCollection $campaignCollection, array $ignoredBuckets, ValidationErrorLogger $errorLogger ) {
+		$this->campaignCollection = $campaignCollection;
+		$this->errorLogger = $errorLogger;
+		$this->ignoredBuckets = $ignoredBuckets;
+	}
+
+	public function isPassing(): bool {
+		return empty( $this->getErrors() );
+	}
+
+	public function getErrors(): array {
+		$this->validate();
+
+		return $this->errorLogger->getErrors();
+	}
+
+	private function validate(): void {
+		if ( $this->hasValidated ) {
+			return;
+		}
+
+		$featureToggleChecks = $this->getFeatureToggleChecks();
+		if ( $duplicates = $this->getDuplicates( $featureToggleChecks ) ) {
+			foreach ( $duplicates as $duplicate ) {
+				$this->errorLogger->addError(
+					'ChoiceFactory contains multiple checks for feature toggle: ' . $duplicate
+				);
+			}
+			return;
+		}
+
+		$configBucketIdList = $this->buildBucketIdsFromCampaignCollection(
+			$this->campaignCollection
+		);
+
+		/** Remove all buckets which were marked as test buckets in TEST_BUCKET_IGNORE_LIST */
+		$featureToggleChecks = array_filter( $featureToggleChecks, [ __CLASS__, 'filterBucketIds' ] );
+		$configBucketIdList = array_filter( $configBucketIdList, [ __CLASS__, 'filterBucketIds' ] );
+
+		/** Check if either the configuration files or the choice factory has rules not found in the other */
+		$codeConfigDiff = array_diff( $featureToggleChecks, $configBucketIdList );
+		$configCodeDiff = array_diff( $configBucketIdList, $featureToggleChecks );
+
+		if ( empty( $codeConfigDiff ) === false || empty( $configCodeDiff ) === false ) {
+			foreach ( $codeConfigDiff as $missingBucket ) {
+				$this->errorLogger->addError(
+					'Bucket ' . $missingBucket . ' is configured but no implementation can be found in ChoiceFactory.'
+				);
+			}
+			foreach ( $configCodeDiff as $featureToggleCheck ) {
+				$this->errorLogger->addError(
+					'Feature toggle check for ' . $featureToggleCheck . ' is implemented but no campaign configuration can be found.'
+				);
+			}
+		}
+
+		$this->hasValidated = true;
+	}
+
+	private function getFeatureToggleChecks(): array {
```

Comment:
This class does too much IMHO: loading/parsing a factory file into bucket IDs **and** comparing the bucket IDs to the list of bucket IDs from the CampaignCollection. Could you introduce a class that only does the parsing and inject it as a dependency into the validator? Name suggestions: `BucketIdDetector`, `BucketIdParser`, etc.

Both classes can then be tested separately and you can inject the test for the validator with a fake.

The test for the parser class could have a constructor with file and method names as arguments (and its own constants as defaults). In the FunFunFactory you'd instantiate it without arguments, in the unit test you could pass in the test file itself and have a dummy method in the test with calls to `featureIsActive`. This way you'll avoid having a seemingly useless PHP file in `Tests\Data`.